README.md
# uba-build

[![npm version](https://img.shields.io/npm/v/uba-build.svg)](https://www.npmjs.com/package/uba-build)
[![devDependency Status](https://img.shields.io/david/dev/tinper-uba/uba-build.svg)](https://david-dm.org/tinper-uba/uba-build#info=devDependencies)
[![NPM downloads](http://img.shields.io/npm/dt/uba-build.svg?style=flat)](https://npmjs.org/package/uba-build)

---

[![NPM](https://nodei.co/npm/uba-build.png)](https://nodei.co/npm/uba-build/)

---

## 介绍

基于 [uba](https://github.com/iuap-design/tinper-uba/) 集成开发工具扩展的插件，可以帮助我们构建uba前端项目。


## 安装


```bash
$ npm install uba -g
```

无需单独安装，安装`uba`工具内置集成。

工具是依赖在开发框架内去使用的，具体参考开发框架里的`package.json`的启动脚本命令，配合`uba.config.js`文件使用


## 参数

下面是在配置script命令传入的参数如下：

```bash
  "scripts": {
    "build": "uba build --noProcess"
  }
```

名称 | 说明
---|---
noProcess | 不显示构建进度


CHANGELOG.md
# CHANGELOG

### version 2.1.1

- 锁定包版本

### version 2.1.0

- 调整部分代码逻辑

### version 1.0.2

- 增加取消查看进度条参数`--noProcess`。

### version 0.0.5

- 更换`webpack 2.7.0`解决`webpack3`内部构建压缩慢的问题。


config.js
const path = require("path");
const CleanWebpackPlugin = require('clean-webpack-plugin');


module.exports = env => {
  let uba = {
    config: {
      plugins: {
        static: {
          root: "static"
        },
        mock: {
          file: "uba.mock.js"
        },
        proxy: [{
          router: "/api/*",
          target: "https://cnodejs.org"
        }, {
          router: ["/users/*", "/orgs/*"],
          target: "https://api.github.com"
        }]
      },
      css: {
        name: "[name].[hash:8].css"
      },
      html: {
        xhtml: true,
        inject: "body",
        hash: env.production,
        filename: 'index.html',
        template: "./view/index.html"
      },
      js: {
        min: env.production,
        optimize: true,
        opt: {
          test: /\.js($|\?)/i
        },
        name: "vendor",
        filename: "[name].[hash:8].js"
      },
      img: {
        limit: 8192,
        name: "images/[name].[hash:8].[ext]"
      }
    },
    pack: {
      devtool: env.production ? "cheap-module-source-map" : "cheap-module-eval-source-map",
      entry: {
        app: env.production ? "./entry" : ["./entry", env.HMR]
      },
      output: {
        path: path.resolve(__dirname, "public"),
        filename: '[name].[hash:8].bundle.js',
        publicPath: "/"
      },
      resolve: {
        extensions: [
          ".json"
        ],
        alias: {
          pages: path.resolve(__dirname, "src/pages/")
        }
      },
      plugins: []
    }
  }
  if (env.production) {
    uba.pack.plugins.push(new CleanWebpackPlugin("public"))
  }
  return uba;
}


src/index.js
/* uba-build
 * @Author: Kvkens(yueming@yonyou.com)
 * @Date:   2017-5-15 00:00:00
 * @Last Modified by:   Kvkens
 * @Last Modified time: 2018-12-29 10:57:08
 */

const chalk = require("chalk");
const webpack = require("webpack");
const util = require("./util");
const webpackConfig = util.getConfig().prodConfig;

/**
 * 获取帮助
 */
function getHelp() {
  console.log(chalk.green(" Usage : "));
  console.log();
  console.log(chalk.green(" uba build"));
  console.log();
  process.exit(0);
}

/**
 * 获取version
 */
function getVersion() {
  console.log(chalk.green(require("../package.json").version));
  process.exit(0);
}

/**
 * 构建资源处理
 */
function build() {
  console.log();
  console.log(chalk.green(`********************************************`));
  console.log(chalk.yellow(` ❤️  uba-build`));
  console.log(chalk.green(` [core] : v${util.getPkg().version}`));
  console.log();
  console.log(chalk.green(` Good luck please wait`));
  console.log(chalk.green(`********************************************`));
  console.log();
  webpack(webpackConfig, function (err, stats) {
    if (!err) {
      console.log('\n' + stats.toString({
        hash: false,
        chunks: false,
        children: false,
        colors: true
      }));
    } else {
      console.log(chalk.red(err));
    }
  });
}

module.exports = {
  plugin: function (options) {
    commands = options.cmd;
    pluginname = options.name;
    if (options.argv.h || options.argv.help) {
      getHelp();
    }
    if (options.argv.v || options.argv.version) {
      getVersion();
    }

    build();

  }
}


src/build.config.js
/**
 * @description develop
 */

const path = require('path');
const chalk = require('chalk');
const glob = require('glob');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const webpack = require('webpack');
const merge = require('webpack-merge');
const argv = require("minimist")(process.argv.slice(2));
const commands = argv;
const util = require('./util');
const base = require('./base.config');
const cfg = util.getUbaConfig()();

//默认的配置用于merge操作
const config = {
    devtool: cfg.devtool ? cfg.devtool : "source-map",
    mode: 'production',
    optimization: {
      minimizer: [
        new UglifyJsPlugin({
          cache: true,
          parallel: true,
          sourceMap: true // set to true if you want JS source maps
        }),
        new OptimizeCSSAssetsPlugin({})
      ]
    },
    output: cfg.output,
    externals: cfg.externals,
    resolve: cfg.resolve,
    module: {
      rules: cfg.loader
    },
    plugins: []
  }

  //传入插件设置
  !commands.noProcess && config.plugins.push(new webpack.ProgressPlugin());
config.plugins = config.plugins.concat(cfg.buildPlugins);

//当前应用模式 单页
if (cfg.appType === 'single') {
  config['entry'] = cfg.entry;
  //设置一次HTML插件
  config['plugins'].push(new HtmlWebpackPlugin(Object.assign({
    template: "./src/index.html",
    minify: {
      removeComments: true,
      collapseWhitespace: true,
      removeRedundantAttributes: true,
      useShortDoctype: true,
      removeEmptyAttributes: true,
      removeStyleLinkTypeAttributes: true,
      keepClosingSlash: true,
      minifyJS: true,
      minifyCSS: true,
      minifyURLs: true
    }
  }, cfg.html)));
} else if (cfg.appType === 'multi') { //多页模式
  let entries = {};
  config['entry'] = {};
  //按照传入glob参数匹配扫描
  glob.sync(cfg.entry).forEach(path => {
    const chunk = path.split("./src/pages/")[1].split("/index.js")[0];
    entries[`${chunk}`] = path;
    let htmlConfig = {
      template: `./src/pages/${chunk}/index.html`,
      chunks: ['manifest', 'vendor', 'commons', chunk],
      chunksSortMode: "manual",
      filename: `${chunk}.html`,
      minify: {
        removeComments: true,
        collapseWhitespace: true,
        removeRedundantAttributes: true,
        useShortDoctype: true,
        removeEmptyAttributes: true,
        removeStyleLinkTypeAttributes: true,
        keepClosingSlash: true,
        minifyJS: true,
        minifyCSS: true,
        minifyURLs: true
      }
    };
    config['plugins'].push(new HtmlWebpackPlugin(Object.assign(htmlConfig, cfg.html)));
  });
  config['entry'] = entries;
}


module.exports = merge(base, config);


src/util.js
/* util
 * @Author: Kvkens(yueming@yonyou.com)
 * @Date:   2017-5-15 00:00:00
 * @Last Modified by:   Kvkens
 * @Last Modified time: 2018-12-29 10:58:41
 */

const path = require("path");
const chalk = require("chalk");
const webpack = require("webpack");
const argv = require("minimist")(process.argv.slice(2));
const commands = argv;

/**
 * 获取uba客户端配置
 */
exports.getConfig = function () {
  let ubaConfig;
  try {
    ubaConfig = require(path.resolve(".", "uba.config.js"));
    !commands.noProcess && ubaConfig.prodConfig.plugins.push(new webpack.ProgressPlugin());
  } catch (e) {
    console.log(chalk.red(e));
    console.log(chalk.yellow("Please check uba.config.js configuration file"));
    process.exit(0);
  }
  return ubaConfig;
}

/**
 * 获取package.json
 */
exports.getPkg = () => {
  return require("../package.json");
}